lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp

```diff
@@ -2715,5 +2739,408 @@ TEST_F(MmeAppProcedureTest, TestTauRejDueToInvalidTac) {
+      nas_msg_auth_resp, sizeof(nas_msg_auth_resp), plmn);
+
+  // Wait for DL NAS Transport for once
+  cv.wait_for(lock, std::chrono::milliseconds(STATE_MAX_WAIT_MS));
+  // Constructing and sending SMC Response to mme_app mimicing S1AP
+  send_mme_app_uplink_data_ind(
+      nas_msg_smc_resp, sizeof(nas_msg_smc_resp), plmn);
+
+  // Sending ULA to mme_app mimicing successful S6A response for ULR
+  send_s6a_ula(imsi, true);
+
+  // Constructing and sending Create Session Response to mme_app mimicing SPGW
+  send_create_session_resp(REQUEST_ACCEPTED, DEFAULT_LBI);
+
+  // Constructing and sending ICS Response to mme_app mimicing S1AP
+  send_ics_response();
+
+  // Constructing UE Capability Indication message to mme_app
+  // mimicing S1AP with dummy radio capabilities
+  send_ue_capabilities_ind();
+
+  // Constructing and sending Attach Complete to mme_app
+  // mimicing S1AP
+  send_mme_app_uplink_data_ind(
+      nas_msg_attach_comp, sizeof(nas_msg_attach_comp), plmn);
+
+  // Wait for DL NAS Transport for EMM Information
+  cv.wait_for(lock, std::chrono::milliseconds(STATE_MAX_WAIT_MS));
+
+  nas_message_t nas_msg_decoded = {0};
+  emm_security_context_t emm_security_context;
+  nas_message_decode_status_t decode_status;
+  status_code_e decoder_rc;
+  decoder_rc = nas_message_decode(
+      nas_msg->data, &nas_msg_decoded, nas_msg->slen,
+      (void*) &emm_security_context, &decode_status);
+  EXPECT_EQ(nas_msg->slen, 67);
+  EXPECT_EQ(decoder_rc, nas_msg->slen);
+  guti = nas_msg_decoded.plain.emm.attach_accept.guti.guti;
+  bdestroy_wrapper(
+      &nas_msg_decoded.plain.emm.attach_accept.esmmessagecontainer);
+  // Destruction at tear down is not sufficient as ICS occurs
+  // twice in this test case.
+  bdestroy_wrapper(&nas_msg);
+
+  // Constructing and sending Modify Bearer Response to mme_app
+  // mimicing SPGW
+  std::vector<int> b_modify = {5};
+  std::vector<int> b_rm     = {};
+  send_modify_bearer_resp(b_modify, b_rm);
+
+  // Check MME state after Modify Bearer Response
+  send_activate_message_to_mme_app();
+  cv.wait_for(lock, std::chrono::milliseconds(STATE_MAX_WAIT_MS));
+  EXPECT_EQ(mme_state_p->nb_ue_attached, 1);
+  EXPECT_EQ(mme_state_p->nb_ue_connected, 1);
+  EXPECT_EQ(mme_state_p->nb_default_eps_bearers, 1);
+  EXPECT_EQ(mme_state_p->nb_ue_idle, 0);
+  EXPECT_EQ(mme_state_p->nb_s1u_bearers, 1);
+
+  // Send PDN connectivity request
+  send_mme_app_uplink_data_ind(
+      nas_msg_pdn_connectivity_req_ims,
+      sizeof(nas_msg_pdn_connectivity_req_ims), plmn);
+
+  ebi_idx++;
+  // Constructing and sending Create Session Response to mme_app mimicing SPGW
+  send_create_session_resp(REQUEST_ACCEPTED, DEFAULT_LBI + ebi_idx);
+
+  // Wait for activate default eps bearer req to be sent
+  cv.wait_for(lock, std::chrono::milliseconds(STATE_MAX_WAIT_MS));
+  cv.wait_for(lock, std::chrono::milliseconds(STATE_MAX_WAIT_MS));
+  cv.wait_for(lock, std::chrono::milliseconds(STATE_MAX_WAIT_MS));
```

Comment:
why are we waiting for 3 message events and for which messages?

Reply:
Waiting for Activate default eps bearer request message. If this time is less, send_erab_setup_rsp is received even before MME sends Activate default eps bearer request message

Reply:
I do not follow. we are waiting for a single default eps bearer request message. Not sure why we need to wait for more here.

Reply:
For every wait call we just wait for 1000ms and not till the message is actually received. Is my understanding correct?

Reply:
It waits at most 1000ms. Thread will resume whichever occurs first: message reception or timeout period specified here by  STATE_MAX_WAIT_MS.

Reply:
If you are expecting to receive only 1 message, you are additionally waiting for 2*STATE_MAX_WAIT_MS unnecessarily.

Reply:
If I just keep 1 wait operation, erab_setup_rsp is getting received before even MME sends Activate default eps bearer request

Reply:
Please check out PR #10747. Sometimes itti message is delayed excessively.

Reply:
I already tested your PR with that change and removed all the unnecessary wait_for calls.

Reply:
Ok let me test with PR #10747 changes

Reply:
Can you share the diff of changes that you did on my PR?

Reply:
unfortunately I did not stash it. But I removed the extra wait_fors you added for `TestNwInitiatedActivateDeactivateDedicatedBearerWithT3485T3495Expirations` at minimum. You should test it by also replacing all these triplet cv.wait_for calls with a single one.